### PR TITLE
Algod: refactor registry operations

### PR DESCRIPTION
**Pull request by algonathan**
## Summary
Refactored WriteThread such that it'll be easier to add new operations. 
Didn't delete the DBs`inner` functions, instead moved them to be handled by a struct that implements `dbOp` interface

### CODE REVIEW

**Review by algoidan — COMMENTED** [review]
looks good!  few minor comments 
Thanks for the changes.

File: data/account/participationRegistry.go [hunk 1]
```diff
@@ -453,38 +452,14 @@ func (db *participationDB) initializeCache() error {
...
+	for op := range db.writeQueue {
+		if err := op.operation.apply(db); err != nil {
 			lastErr = err
 		}
 
-		if wr.flushResultChannel != nil {
-			wr.flushResultChannel <- lastErr
-			lastErr = nil
+		if op.flush != nil {
+			op.flush <- lastErr
```
[the remark this answers is not in the document]
Comment (algonathan) [reply data/account/participationRegistry.go:462]: Good catch! I've added a test to inspect that too.

File: data/account/participationRegistry.go [hunk 2]
```diff
@@ -22,17 +22,15 @@ import (
...
 
-	"github.com/algorand/go-deadlock"
-
 	"github.com/algorand/go-algorand/crypto"
 	"github.com/algorand/go-algorand/crypto/merklesignature"
 	"github.com/algorand/go-algorand/data/basics"
 	"github.com/algorand/go-algorand/logging"
 	"github.com/algorand/go-algorand/protocol"
 	"github.com/algorand/go-algorand/util/db"
+	"github.com/algorand/go-deadlock"
```
Comment (tsachiherman) [inline data/account/participationRegistry.go:33]: this line should remain in it's own group since it's not in the go-algorand repo.

File: go.mod [hunk 3]
```diff
@@ -45,7 +45,8 @@ require (
 	github.com/spf13/pflag v1.0.5 // indirect
 	github.com/stretchr/testify v1.7.0
 	golang.org/x/crypto v0.0.0-20210921155107-089bfa567519
-	golang.org/x/sync v0.0.0-20200625203802-6e8e738ad208 // indirect
+	golang.org/x/net v0.0.0-20210405180319-a5a99cb37ef4 // indirect
+	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c // indirect
```
Comment (tsachiherman) [inline go.mod:49]: unless this PR requires that, I'd rather you'd avoid updating the go.mod/go.sum with benign updates.
( there is minor risk in doing that, and unless it's needed, I'd rather avoid that )

**Review by algoidan — APPROVED** [review]

**Review by winder — DISMISSED** [review]
Only one small suggestion. Nice improvement @algonathan !

File: data/account/registeryDbOps.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	"fmt"
+	"github.com/algorand/go-algorand/protocol"
+	"strings"
+)
+
+type dbOp interface {
+	apply(d *participationDB) error
+}
+
+type opRequest struct {
```
[the remark this answers is not in the document]
Comment (algoidan) [reply data/account/registeryDbOps.go:32]: thanks for the advice 
for to make sure we understand it
the creation of the operation would be:
`db.writeQueue <- makeOpRequest(&appendKeysOp{
			id:   id,
			keys: keys,
		})
	}`
or something similar ?

**Review by winder — APPROVED** [review]

**Review by algoidan — APPROVED** [review]
looks good!

**Pull request merged**